features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,263 @@
 
-When one of events is triggered, skygear sends a HTTP POST with a payload to the webhook's configured URL.
+# Proposed Designs
 
-## SYNC and ASYNC hooks
+## Events
+Web-hook events are triggered when some mutating operation is performed by
+auth gear.
 
-There are two types of webhook:
+Each operation will trigger two events: BEFORE and AFTER:
+- BEFORE events would be triggered before the operation is performed; the
+  operation can be aborted by web-hook handler.
+- AFTER events would be trigger after the operation is performed.
 
-- `SYNC` - gear's operation will be blocked until a response is received. Gear should respect `SYNC` hook's response, if it indicates failed, operation should abort and return a formatted error response to the client.
-- `ASYNC` - gear's operation won't be blocked and the response from the webhook is ignored.
+Both events has the same event payload. Web-hook event handlers cannot modify
+the payload.
 
-## Request Headers
 
-Following headers will be added to webhook POST request:
+## Delivery
 
-- `X-Skygear-Webhook-Req-ID`: an ID of the request.
-- `X-Skygear-Webhook-Async`: if is a `async` hook, then the value is `TRUE` otherwise is `FALSE`. 
-- `X-Skygear-Webhook-Signature`: a signature string of the webhook event, a develop should can verify the request by skygear's SDK or manually.
+The events would be delivered to web-hook handlers as an HTTP POST request to
+the handler endpoint.
 
-## Request Payload
+The handler endpoint can be specified as full URL or an absolute path. Absolute
+paths will be resolved to full URL using an inferred URL scheme and authority:
+- from HTTP request URL if the event is generated from it; otherwise,
+- from the tenant configuration.
 
-Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+Each event type can have multiple handlers; the order of deliveries is
+unspecified.
 
-All event payloads follows following payload format:
+BEFORE events will always be delivered before AFTER events. BEFORE events will
+be delivered in well-defined order during a request. AFTER events will be
+delivered in unspecified order.
 
-```json=
+Web-hook handler must be idempotent.
+
+Web-hook handler must return a status code within the 2XX range. Responding with
+status code outside the range, including 3XX & 5XX, would be considered a
+failed delivery.
+
+The time spent in a delivery must not exceed 5 seconds, otherwise would be
+considered a failed delivery.
+
+### BEFORE Events
+
+BEFORE events would be delivered to web-hook handlers synchronously.
+
+Web-hook handler should respond with a JSON-formatted body to indicated whether
+the operation should be failed, for example:
+```json
+// Allowed
 {
-    "event": "event_name",
+    "is_allowed": true
+}
+
+// Disallowed
+{
+    "is_allowed": false,
+    "reason": "the metadata does not match the required format.",
     "data": {
```

Comment:
Will our hook spec include the data return value?

In the old spec, there was a very implicit assumption (which I think is not so good) -- after the hook is completed, the hook can return the DB object (say the user object) for the updated value.

It is useful for cases like:
* if we don't auto-populate the avatar from Facebook Login, a user may want to write a hook that populates the metadata with an avatar from Facebook.
* with some crazy manual condition, like any email address, ends at @oursky.com, the user is always verified or disabled by default -- whatever crazy rules users want to write.

It is helpful as it:
1. Avoid calling more APIs to modify the data, hence lower the call roundtrip time and improve the performance of each webhook
2. Avoid potential recursive lock user may create

Do we want to include the same spec for the return data of auth hooks? That they can update some attributes?

Reply:
This data is just the mentioned 'additional information' for the disallow action; it will be returned in the HTTP response error body.

I don't plan to support modifying the request using web-hook handler.

One of my concern is the audit feature. If web-hook can modify the request arbitrarily, audit logs can no longer be relied on to reflect real action of users.

Reply:
However, if we actually specify what data could be modified for each hook, aren't we can include that in audit log too?

Reply:
BTW, I think another advantage of having a return value to update some data is:

1. If we don't have it, in any case the hooks need to update, say metadata, it will have some side effect.
2. If the before sync hook is failed, the users need to take care of the inconsistency of side effect.

If we restore the feature of updating some common data to them, it greatly reduce the complexity as the users can rely on Skygear to update some common data only if the hook is being called and returned successfully.

Reply:
If we specify the fields that can be modified (e.g. metadata in before user create, but not password), the same problem remains: we cannot sure the performed operation is exactly as user intended.

If we include the modified/original fields in audit log, some problems may be considered:
- Audit log is supposed to be immutable and keep for a long time, personal information should not be included.

Reply:
I don't get it, logically speaking, no matter it is login hook call update metadata, or login hook return metadata, we could still make an audit log with two entries: Login, and Metadata Updated.

Whether to include those metadata in logs or not is another decision.

Reply:
Agree that we can a separate entry to reflect the fact that the before hook modified part of the payload.

Reply:
Whether updating metadata should produce an audit log entry is another thing to specified in #340, but we can put it aside for now.

After discussing with @louischan-oursky, I'll allow mutation on some limited fields in before sync event handler.

Reply:
Sure. BTW, I think it is better if all auth gear have the same set of fields that users can update for better developer experiences.